cmd/protoc-gen-go-aip/internal/genaip/resourcename.go

```diff
@@ -0,0 +1,331 @@
+package genaip
+
+import (
+	"strconv"
+	"strings"
+
+	"github.com/stoewer/go-strcase"
+	"go.einride.tech/aip/reflect/aipreflect"
+	"go.einride.tech/aip/resourcename"
+	"google.golang.org/genproto/googleapis/api/annotations"
+	"google.golang.org/protobuf/compiler/protogen"
+	"google.golang.org/protobuf/reflect/protoregistry"
+)
+
+type resourceNameCodeGenerator struct {
+	resource *annotations.ResourceDescriptor
+	file     *protogen.File
+	files    *protoregistry.Files
+}
+
+func (r resourceNameCodeGenerator) GenerateCode(g *protogen.GeneratedFile) error {
+	if len(r.resource.Pattern) == 0 {
+		return nil
+	}
+	hasMultiPattern := len(r.resource.Pattern) > 1
+	hasFutureMultiPattern := r.resource.History == annotations.ResourceDescriptor_FUTURE_MULTI_PATTERN
+	// Generate multi-pattern interface and parse methods if we have multiple patterns now or in the future.
+	if hasMultiPattern || hasFutureMultiPattern {
+		if err := r.generateMultiPatternInterface(g); err != nil {
+			return err
+		}
+		if err := r.generateMultiPatternParseMethod(g); err != nil {
+			return err
+		}
+	}
+	// Generate the single-pattern struct unless we explicitly only want multi-patterns from the start.
+	if !hasFutureMultiPattern {
+		if err := r.generatePatternStruct(
+			g, r.resource.Pattern[0], r.SinglePatternStructName(),
+		); err != nil {
+			return err
+		}
+	}
+	// Generate the multi-pattern variant of the single-pattern struct if we need multi-pattern support.
+	if hasMultiPattern || hasFutureMultiPattern {
+		if err := r.generatePatternStruct(
+			g, r.resource.Pattern[0], r.MultiPatternStructName(r.resource.Pattern[0]),
+		); err != nil {
+			return err
+		}
+	}
+	// Generate multi-pattern structs for all but the first pattern.
+	for _, pattern := range r.resource.Pattern[1:] {
+		if err := r.generatePatternStruct(g, pattern, r.MultiPatternStructName(pattern)); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r resourceNameCodeGenerator) generatePatternStruct(
+	g *protogen.GeneratedFile,
+	pattern string,
+	typeName string,
+) error {
+	g.P()
+	g.P("type ", typeName, " struct {")
+	var sc resourcename.Scanner
+	sc.Init(pattern)
+	for sc.Scan() {
+		if sc.Segment().IsVariable() {
+			g.P(strcase.UpperCamelCase(string(sc.Segment().Literal())), " string")
+		}
+	}
+	g.P("}")
+	if err := r.generateValidateMethod(g, pattern, typeName); err != nil {
+		return err
+	}
+	if err := r.generateContainsWildcardMethod(g, pattern, typeName); err != nil {
+		return err
+	}
+	if err := r.generateStringMethod(g, pattern, typeName); err != nil {
+		return err
+	}
+	if err := r.generateMarshalStringMethod(g, typeName); err != nil {
+		return err
+	}
+	if err := r.generateUnmarshalStringMethod(g, pattern, typeName); err != nil {
+		return err
+	}
```

Comment:
looks like most of these never returns errors, so might as well remove error return

Reply:
hmm good point, I'll look into that as a follow up